lib/async-emitter.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+const { iter } = require('@metarhia/common');
+
+class AsyncEmitter {
+  constructor() {
+    this.events = new Map();
+    this.wrappers = new Map();
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  on(name, fn) {
+    let event = this.events.get(name);
+    if (!event) {
+      event = new Set();
+      this.events.set(name, event);
+    }
+    event.add(fn);
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  // Returns: <Promise> | <undefined>
+  once(name, fn) {
+    if (fn === undefined) {
+      return new Promise(resolve => {
+        this.once(name, resolve);
+      });
+    }
+    const wrapper = (...args) => {
+      this.remove(name, fn);
+      return fn(...args);
+    };
+    this.wrappers.set(fn, wrapper);
+    this.on(name, wrapper);
+    return undefined;
+  }
+
+  // Emit event
+  //   name <string> event name
+  //   args <any[]>
+  // Returns: <Promise> | <undefined>
+  async emit(name, ...args) {
+    const event = this.events.get(name);
+    if (!event) return undefined;
+    const listeners = event.values();
+    const promises = iter(listeners).map(fn => fn(...args));
+    return Promise.all(promises);
+  }
+
+  // Remove event listener
+  //   name <string> event name
+  //   fn <Function> listener to remove
+  remove(name, fn) {
+    const { events, wrappers } = this;
+    const event = events.get(name);
+    if (!event) return;
+    if (event.has(fn)) event.delete(fn);
+    const wrapper = wrappers.get(fn);
```

Comment:
What about the case:
```js
ee.on('a', fn);
ee.once('ab', fn);

ee.remove('a', fn);  // once listener will also be removed.
```

Reply:
@nechaido yes, I think it is normal behavior.

Reply:
No, it is definitely not, `once` is for a completely different event.

Reply:
Sorry, yes, it's wrong behavior, we need such case in tests :+1: